### ISSUE TITLE
AIX JDK15+ java.lang.System. mapLibraryName(libname) returns library names with .so suffix

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse/openj9-docs#561]
**Issue or pull request number:**
https://github.com/eclipse/openj9/pull/9489

**Overview:**
`java.lang.System.mapLibraryName(libname)` returns a platform-specific string representing a native library with `.so` suffix at `AIX JDK15+`.
`AIX JDK8/11` still returns a library name with `.a` suffix.
Previously this returns `.a` suffix for all `AIX Java` versions.

A doc PR is to be opened to update the externals / what's new info.

**Release target:**
*Eclipse OpenJ9 0.22.0*

**Applies to the following JDK versions:**
*15 and later*

**Applies to the following platforms:**
*AIX*

**For new command line options:**

**Comment by DanHeidinga** [comment]
@zl-wang We've made this change for 11+ and it would be good to get confirmation from someone with more knowledge of P that this won't cause user problems.

Our understanding is that AIX is moving from `.a` to `.so` for newer libraries.

**Comment by zl-wang** [comment]
@DanHeidinga     no, that is not correct information.

Nothing is changed or intended to be changed with regard to base AIX libraries:  at file-system level, 32-bit and 64-bit (same logical) libraries are still packaged in archive, i.e.  libSomething.a (which can include 32bit shr.o and libSomething.so.x, and 64bit shr_64.so and/or libSomething.so.x members).  The linker only picks up the none-LOADONLY member for compile-time, while the loader can automatically pick up whichever one to run.  i.e.  as it is (has been for 20 years), both shr.o and libSomething.so.x member are supported in archive. Usually, the shr.o member is marked as LOADONLY.  You will not see any explicit libSomething.so.x file for AIX-carried libraries.

For 3rd-party libraries (e.g.  Linux toolbox support, under /opt/freeware), the 3-rd party typically tried to cover all grounds.  They will ship both explicit libSomething.so.x file (to be compatible with Linux I guessed) and libSomething.a archive containing libSomething.so.x member.

All in all, nothing is or has changed.   I am not quite sure how this topic came up.  Maybe you are misled by /opt/freeware look-up (there are both .so and .a there). dlopen can use either .o or .so member from an archive at runtime.

**Comment by DanHeidinga** [comment]
@JasonFengJ9 Do you have a reference for the `.a` -> `.so` change?  On the basis of Julian's comments, we may want to undo eclipse/openj9#9489

**Comment by JasonFengJ9** [comment]
The change `.a` -> `.so` is required by JDK15 at least, without it no native library can be loaded in AIX platform.
The code in question is https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/521b2f12fc414d54b42baebf0cbdec6ec8a27390/src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java#L307-L323

```
    private NativeLibrary findFromPaths(String[] paths, Class<?> fromClass, String name) {
        for (String path : paths) {
            File libfile = new File(path, System.mapLibraryName(name)); --> get platform specific native library name, only expect .so, .a won't work in later loadLibrary()
            NativeLibrary nl = loadLibrary(fromClass, libfile);
            if (nl != null) {
                return nl;
            }
            libfile = ClassLoaderHelper.mapAlternativeName(libfile);
            if (libfile != null) {
                nl = loadLibrary(fromClass, libfile);
                if (nl != null) {
                    return nl;
                }
            }
        }
        return null;
    }
```

It appears `ClassLoaderHelper.mapAlternativeName(libfile)` need add another case for `AIX` besides existing `OSX .jnilib` case at https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/81a6e58c25e4ce880037d14345a068011e4a0d3f/src/java.base/macosx/classes/jdk/internal/loader/ClassLoaderHelper.java#L35-L48.
```
    /**
     * Returns an alternate path name for the given file
     * such that if the original pathname did not exist, then the
     * file may be located at the alternate location.
     * For mac, this replaces the final .dylib suffix with .jnilib
     */
    static File mapAlternativeName(File lib) {
        String name = lib.toString();
        int index = name.lastIndexOf('.');
        if (index < 0) {
            return null;
        }
        return new File(name.substring(0, index) + ".jnilib");
    }
```

**Comment by JasonFengJ9** [comment]
@zl-wang the issue we had is regarding `java.lang.System.mapLibraryNamepublic(libname)` which maps a library name into a platform-specific string representing a native library, i.e., it returns a name which has  either `.a` or `.so` suffix.
Since both format are valid in AIX, we would like to return one format preferred by other applications. All `AIX JDK8` internal libraries are in `.so` format. How's the trend for other AIX applications particularly those `JNI/JVMTI` native libraries?
In short, is there a preferred (or more popular) application library format, archive libraries (.a) vs shared objects (.so)?

**Comment by zl-wang** [comment]
@JasonFengJ9  the support for all .a (archive package for both 32/64bit)/.o (executable)/.so (executable) has been there for almost 20 years. nothing in this regard has changed.  explicit .so files so far only have existed for 3-rd party software (freeware or IBM JDK, e.g.) ... i.e.  base AIX always shipped libraries in .a archive(s).  i don't think we should change things in this regard, comparing to java8 or java11.  it appears to me that java15 missed AIX compatibility somehow, and it should be worked back into.

**Comment by JasonFengJ9** [comment]
> explicit .so files so far only have existed for 3-rd party software (freeware or IBM JDK, e.g.) ... 

@zl-wang thanks the insights. Does the comment above indicate that the 3rd party software like `IBM JDK` and Java native applications (`JNI/JVMTI`) are adopting `.so` format more than `.a`?

**Comment by zl-wang** [comment]
@JasonFengJ9  it appeared to me not an adopting issue. After all, .a will contain .so or .o executable member(s) anyway.  it looks more like a practical issue and the original vision (single .a for both 32/64 bit easier management) didn't pan out in practice on other OS(es).  Linux went the way with explicit lib32 and lib64 directories. AIX kept this way of doing things. I don't think there are substantial oppositions to shipping .so libraries on AIX, but java15 should not expect .a lib to disappear suddenly. ultimately, as long as java15 is made to work, not a problem at all.

**Comment by JasonFengJ9** [comment]
> java15 should not expect .a lib to disappear suddenly.

Agreed, that's why OpenJDK hotspot fails some tests containing `.a` libraries within a jar file.
I am proposing JCL change to search for alternative library names via `jdk.internal.loader.ClassLoaderHelper.mapAlternativeName(File lib)`, more details are at https://github.com/eclipse/openj9/issues/9518.

**Comment by JasonFengJ9** [comment]
> Do you have a reference for the .a -> .so change? On the basis of Julian's comments, we may want to undo eclipse/openj9#9489

@DanHeidinga Apology for late response, I still can't find the official reference I thought the trend switching from `.a` to `.so`, it probably was an impression from an informal forum user comment but without official links. Sorry for the confusion this might have caused.
I have opened https://github.com/eclipse/openj9/issues/9518 for further discussion if https://github.com/eclipse/openj9/pull/9489 should be reverted.

**Comment by pmhayward** [comment]
@JasonFengJ9  (cc @DanHeidinga) ... Hi Jason... can you confirm the status of this doc issue, please. Does someone need to do some work to get it into the 0.21.0 docs, or is it postponed/shelved/cancelled?!

**Comment by JasonFengJ9** [comment]
@pmhayward thanks the reminder, no action required for this issue yet.

There is no change for pre-JDK15 (i.e., 0.21.0 release).
JDK15+ behaviour has been changed, i.e., `.so` suffix is returned from `System. mapLibraryNamepublic(libname)` though it is still being discussed at https://github.com/eclipse/openj9/issues/9788.

**Comment by pmhayward** [comment]
Thanks - will move this issue to 0.22.0

**Comment by pmhayward** [comment]
@JasonFengJ9  ... Hi Jason, Just keeping tabs on outstanding 0.22.0 doc issues. Is this one still in plan for this release? FYI, we're aiming to get all docs finished by milestone 2 (31 Aug), in part due to vacations in our team. Thanks!

**Comment by JasonFengJ9** [comment]
@pmhayward Yes, it is. This change has been applied to `JDK15+`, and probably is going to stay since it appears `OpenJDK` community will keep it as it is as per [1].

[1] https://github.com/eclipse/openj9/issues/9788#issuecomment-654292377

**Comment by pmhayward** [comment]
@JasonFengJ9 Thanks for the update ... hopefully you'll be able to get a PR done for the doc updates in the next couple of weeks.

**Comment by JasonFengJ9** [comment]
Doc PR is created at https://github.com/eclipse/openj9-docs/pull/620.

**Comment by SueChaplain** [comment]
Further fix by https://github.com/eclipse/openj9-docs/pull/625 - closing
